### PR TITLE
Make the tests faster

### PR DESCRIPTION
**Pull request by calebmer**
Some of the test suites are pretty slow, so this PR aims to speed up our tests. With this PR tests go from 42 seconds to 12 seconds on my machine 🎉

The secret to pretty much all of the perf wins is parallelization. Jest does not automatically parallelize tests which is fine when testing synchronous code, but is not ok for some of the heavy asynchronous tests we end up running.

We were also repeating a lot of introspection work so I tried to get clients and introspections to be shared as much as possible.

### CODE REVIEW

**Comment by calebmer** [comment]
It looks like tests on CI went from 30s to 10s.

**Comment by benjie** [comment]
Nice. Looks like we can't use snapshot tests when we're expecting errors though:

```
 FAIL  src/postgraphql/__tests__/postgraphqlIntegrationQueries-test.js
  ● dynamic-json.graphql
    expect(value).toMatchSnapshot()
    
    Received value does not match stored snapshot 1.
    
    - Snapshot
    + Received
    
    @@ -1,5 +1,5 @@
     Object {
       "errors": Array [
    -    [SyntaxError: Unexpected token h in JSON at position 0],
    +    [SyntaxError: Unexpected token h],
       ],
     }
```

File: src/__tests__/utils/createTestInParallel.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    testFns.push(fn)
+
+    test(name, async () => {
+      // If the tests have not yet been run then run all of our tests.
+      if (!testResults) {
+        testResults = testFns.map(testFn => Promise.resolve(testFn()))
+      }
+      // Await the result.
+      await testResults[index]
+    })
```
Comment (benjie) [inline src/__tests__/utils/createTestInParallel.ts]: Nice.

**Comment by calebmer** [comment]
Snapshot tests are still fine. That one snapshot test failed because it looks like the error message is different in Node.js 4 then in Node.js 6, however seeing that helped me catch a bug I forgot to fix 😊

The second error causing the tests to crash is because Koa ships some ES2015 syntax in its source code that Node.js 4 doesn’t like. So the fix was to not test Koa if the Node.js major version is less than or equal to 4.

**Pull request merged**